File: zig/src/core/freqdist.zig
const std = @import("std");
const ascii = @import("ascii.zig");
const types = @import("types.zig");

pub fn updateCount(map: *std.AutoHashMap(u64, u64), key: u64) types.CountError!void {
    if (map.getPtr(key)) |count| {
        count.* += 1;
        return;
    }
    map.put(key, 1) catch return error.OutOfMemory;
}

pub fn updateCountU128(map: *std.AutoHashMap(u128, u64), key: u128) types.CountError!void {
    if (map.getPtr(key)) |count| {
        count.* += 1;
        return;
    }
    map.put(key, 1) catch return error.OutOfMemory;
}

pub fn buildTokenFreqMapAscii(input: []const u8, allocator: std.mem.Allocator) types.CountError!std.AutoHashMap(u64, u64) {
    var map = std.AutoHashMap(u64, u64).init(allocator);
    errdefer map.deinit();

    var in_token = false;
    var token_hash: u64 = ascii.FNV_OFFSET_BASIS;

    for (input) |ch| {
        if (ascii.isTokenChar(ch)) {
            if (!in_token) {
                in_token = true;
                token_hash = ascii.FNV_OFFSET_BASIS;
            }
            token_hash = ascii.tokenHashUpdate(token_hash, ch);
        } else if (in_token) {
            try updateCount(&map, token_hash);
            in_token = false;
        }
    }

    if (in_token) {
        try updateCount(&map, token_hash);
    }

    return map;
}

pub fn buildNgramFreqMapAscii(input: []const u8, n: usize, allocator: std.mem.Allocator) types.CountError!std.AutoHashMap(u64, u64) {
    if (n == 0) return error.InvalidN;

    var map = std.AutoHashMap(u64, u64).init(allocator);
    errdefer map.deinit();

    const window = allocator.alloc(u64, n) catch return error.OutOfMemory;
    defer allocator.free(window);

    var seen_tokens: usize = 0;
    var in_token = false;
    var token_hash: u64 = ascii.FNV_OFFSET_BASIS;

    for (input) |ch| {
        if (ascii.isTokenChar(ch)) {
            if (!in_token) {
                in_token = true;
                token_hash = ascii.FNV_OFFSET_BASIS;
            }
            token_hash = ascii.tokenHashUpdate(token_hash, ch);
        } else if (in_token) {
            window[seen_tokens % n] = token_hash;
            seen_tokens += 1;
            if (seen_tokens >= n) {
                const start = (seen_tokens - n) % n;
                try updateCount(&map, ascii.hashNgram(window, start, n));
            }
            in_token = false;
        }
    }

    if (in_token) {
        window[seen_tokens % n] = token_hash;
        seen_tokens += 1;
        if (seen_tokens >= n) {
            const start = (seen_tokens - n) % n;
            try updateCount(&map, ascii.hashNgram(window, start, n));
        }
    }

    return map;
}

pub fn fillFromMap(map: *std.AutoHashMap(u64, u64), out_hashes: []u64, out_counts: []u64) types.CountError!void {
    if (out_hashes.len != out_counts.len) return error.InsufficientCapacity;

    const unique = map.count();
    if (out_hashes.len < unique) return error.InsufficientCapacity;

    var idx: usize = 0;
    var iter = map.iterator();
    while (iter.next()) |entry| {
        out_hashes[idx] = entry.key_ptr.*;
        out_counts[idx] = entry.value_ptr.*;
        idx += 1;
    }
}

test "freqdist building" {
    const allocator = std.testing.allocator;
    const input = "this this is is a a test test";

    var token_map = try buildTokenFreqMapAscii(input, allocator);
    defer token_map.deinit();
    try std.testing.expectEqual(@as(usize, 4), token_map.count());

    var ngram_map = try buildNgramFreqMapAscii(input, 2, allocator);
    defer ngram_map.deinit();
    try std.testing.expectEqual(@as(usize, 7), ngram_map.count());
}

test "freqdist invalid n" {
    const allocator = std.testing.allocator;
    const result = buildNgramFreqMapAscii("abc", 0, allocator);
    try std.testing.expectError(error.InvalidN, result);
}

File: rust/src/freqdist.rs
use crate::ascii;
use crate::CoreResult;
use std::collections::HashMap;

pub fn update_count(map: &mut HashMap<u64, u64>, key: u64) {
    *map.entry(key).or_insert(0) += 1;
}

pub fn update_count_u128(map: &mut HashMap<u128, u64>, key: u128) {
    *map.entry(key).or_insert(0) += 1;
}

pub fn build_token_freq_map_ascii(input: &[u8]) -> CoreResult<HashMap<u64, u64>> {
    let mut map: HashMap<u64, u64> = HashMap::new();

    let mut in_token = false;
    let mut token_hash = ascii::FNV_OFFSET_BASIS;

    for &ch in input {
        if ascii::is_token_char(ch) {
            if !in_token {
                in_token = true;
                token_hash = ascii::FNV_OFFSET_BASIS;
            }
            token_hash = ascii::token_hash_update(token_hash, ch);
        } else if in_token {
            update_count(&mut map, token_hash);
            in_token = false;
        }
    }

    if in_token {
        update_count(&mut map, token_hash);
    }

    Ok(map)
}

pub fn build_ngram_freq_map_ascii(input: &[u8], n: usize) -> CoreResult<HashMap<u64, u64>> {
    if n == 0 {
        return Err(crate::CoreError::InvalidN);
    }

    let mut map: HashMap<u64, u64> = HashMap::new();
    let mut window = vec![0u64; n];

    let mut seen_tokens: usize = 0;
    let mut in_token = false;
    let mut token_hash = ascii::FNV_OFFSET_BASIS;

    for &ch in input {
        if ascii::is_token_char(ch) {
            if !in_token {
                in_token = true;
                token_hash = ascii::FNV_OFFSET_BASIS;
            }
            token_hash = ascii::token_hash_update(token_hash, ch);
        } else if in_token {
            window[seen_tokens % n] = token_hash;
            seen_tokens += 1;
            if seen_tokens >= n {
                let start = (seen_tokens - n) % n;
                update_count(&mut map, ascii::hash_ngram(&window, start, n));
            }
            in_token = false;
        }
    }

    if in_token {
        window[seen_tokens % n] = token_hash;
        seen_tokens += 1;
        if seen_tokens >= n {
            let start = (seen_tokens - n) % n;
            update_count(&mut map, ascii::hash_ngram(&window, start, n));
        }
    }

    Ok(map)
}

pub fn fill_from_map(
    map: &HashMap<u64, u64>,
    out_hashes: &mut [u64],
    out_counts: &mut [u64],
) -> CoreResult<()> {
    if out_hashes.len() != out_counts.len() {
        return Err(crate::CoreError::InsufficientCapacity);
    }

    let unique = map.len();
    if out_hashes.len() < unique {
        return Err(crate::CoreError::InsufficientCapacity);
    }

    let mut idx: usize = 0;
    for (key, value) in map.iter() {
        out_hashes[idx] = *key;
        out_counts[idx] = *value;
        idx += 1;
    }

    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn freqdist_building() {
        let input = b"this this is is a a test test";

        let token_map = build_token_freq_map_ascii(input).unwrap();
        assert_eq!(token_map.len(), 4);

        let ngram_map = build_ngram_freq_map_ascii(input, 2).unwrap();
        assert_eq!(ngram_map.len(), 7);
    }

    #[test]
    fn freqdist_invalid_n() {
        let result = build_ngram_freq_map_ascii(b"abc", 0);
        assert!(result.is_err());
    }
}
